Make more stock classes configurable

jarednorman opened the pull request:
**Description**

This makes three additional stock classes configurable, so that they can be overridden by extensions like `solidus_product_assembly` or any stores that need similar functionality:

- `Spree::Stock::InventoryUnitBuilder`
- `Spree::Stock::InventoryValidator`
- `Spree::Stock::AvailabilityValidator`

I think there's probably more (breaking) work to be done to improve how this system works and better support the use case of having line items that map to inventory units with different quantities or variants, but this is a good first step.

**Checklist:**
- [x] I have followed [Pull Request guidelines](https://github.com/solidusio/solidus/blob/master/CONTRIBUTING.md#pull-request-guidelines)
- [x] I have added a detailed description into each commit message
- [x] I have updated Guides and README accordingly to this change (if needed)
- [x] I have added tests to cover this change (if needed)
- [x] I have attached screenshots to this PR for visual changes (if needed)

spaghetticode reviewed (approved):
Great change, thanks @jarednorman !

kennyadsl reviewed (approved):
Thanks Jared! Approved and left some suggestions to try (if not done already) to avoid stubbing the preference manually. Let me know if those are valid or if I am missing something.

jarednorman commented on core/spec/models/spree/order_spec.rb:
```diff
@@ -1600,6 +1600,42 @@ def generate
+        # other tests.
+        allow(Spree::Config.stock)
+          .to receive(:inventory_unit_builder_class)
+          .and_return(TestInventoryUnitBuilder)
```
I'll give that a shot.

jarednorman replied:
Thanks @kennyadsl, that worked.

jarednorman commented:
Only change here is that I applied @kennyadsl's suggestion for stubbing the preferences.

waiting-for-dev commented on core/spec/lib/spree/core/stock_configuration_spec.rb:
```diff
@@ -92,4 +92,26 @@
+    end
+
+    context "with another constant name assigned" do
+      MyInventoryUnitBuilder = Class.new
```
It has the side effect of leaving a top-level defined constant for the rest of the test suite. What about removing it with `Object.send(:remove_const, :MyInventoryUnitBuilder)`?

waiting-for-dev commented on core/spec/models/spree/order_spec.rb:
```diff
@@ -1600,6 +1600,40 @@ def generate
+
+    context "with a custom inventory unit builder" do
+      before do
+        class TestInventoryUnitBuilder
```
Like before, we can make sure we remove the defined constant.

waiting-for-dev commented on core/spec/models/spree/order_spec.rb:
```diff
@@ -1600,6 +1600,40 @@ def generate
+        # arbitrary_inventory_unit let variable.
+        allow_any_instance_of(TestInventoryUnitBuilder)
+          .to receive(:missing_units_for_line_item)
+          .and_return([arbitrary_inventory_unit])
```
I think you can work around this issue if you define the class like the following:

```ruby
TestInventoryUnitBuilder = Class.new do
  def initialize(order)
  end

  define_method(:missing_units_for_line_item) do |_line_item|
    [arbitrary_inventory_unit]
  end
end
```

jarednorman replied:
I thought I tried that and it didn't work... but let me double check that.

jarednorman replied:
Ah, I just needed to capture it in a local variable.

waiting-for-dev commented on core/spec/models/spree/order_spec.rb:
```diff
@@ -1600,6 +1600,40 @@ def generate
+        test_stock_config.inventory_unit_builder_class = TestInventoryUnitBuilder.to_s
+        stub_spree_preferences(stock: test_stock_config)
+
+        order.shipments.first.update!(created_at: 2.years.ago)
```
What about creating a targeted order (instead of the one defined "globally")? It'd make intentions more apparent, and we wouldn't need this workaround to be sure we select the correct shipment.

jarednorman replied:
I'm not sure what you're suggesting here. Could you clarify?

waiting-for-dev replied:
Hey @jarednorman, I'm still confused about having to set `2.years.ago` here :thinking: 

Wouldn't it be possible something like that?

```ruby
# A single shipment with no inventory unit
let(:order) { create(:order, shipments: create(:shipment, inventory_units: [])) }
before do
 # Everything except the "order.shipments.first.update!(created_at: 2.years.ago)" line
end
# ...
it "relies on the custom builder" do
# replacing the second except with:
expect(order.shipments.first.inventory_units).to contain_exactly arbitrary_inventory_unit
end
```

waiting-for-dev reviewed (commented):
Thanks, @jarednorman. It's always nice having more flexibility. I left some suggestions.

jarednorman commented:
I've updated the description and added two additional commits to hook into the other two spots where `spree_product_assembly` (which I'm not actually using... but it's a good reference) overrides all the methods on an object.

waiting-for-dev commented on core/spec/lib/spree/core/stock_configuration_spec.rb:
```diff
@@ -3,93 +3,145 @@
-      it "returns the constant" do
-        is_expected.to be MyCoordinator
-      end
+      Object.send(:remove_const, :MyCoordinator)
```
```suggestion
      ensure
        Object.send(:remove_const, :MyCoordinator)
```

waiting-for-dev commented on core/spec/lib/spree/core/stock_configuration_spec.rb:
```diff
@@ -3,93 +3,145 @@
-      it "returns the constant" do
-        is_expected.to be MyEstimator
-      end
+      Object.send(:remove_const, :MyEstimator)
```
```suggestion
      ensure
        Object.send(:remove_const, :MyEstimator)
```

waiting-for-dev commented on core/spec/lib/spree/core/stock_configuration_spec.rb:
```diff
@@ -3,93 +3,145 @@
-      end
+      expect(subject).to be MySorter
+
+      Object.send(:remove_const, :MySorter)
```
```suggestion
      ensure
        Object.send(:remove_const, :MySorter)
```

waiting-for-dev commented on core/spec/lib/spree/core/stock_configuration_spec.rb:
```diff
@@ -3,93 +3,145 @@
+
+      expect(subject).to be MyInventoryUnitBuilder
+
+      Object.send(:remove_const, :MyInventoryUnitBuilder)
```
```suggestion
      ensure
        Object.send(:remove_const, :MyInventoryUnitBuilder)
```

waiting-for-dev commented on core/spec/lib/spree/core/stock_configuration_spec.rb:
```diff
@@ -3,93 +3,145 @@
+
+      expect(subject).to be MyAvailabilityValidator
+
+      Object.send(:remove_const, :MyAvailabilityValidator)
```
```suggestion
    ensure
      Object.send(:remove_const, :MyAvailabilityValidator)
```

waiting-for-dev commented on core/spec/lib/spree/core/stock_configuration_spec.rb:
```diff
@@ -3,93 +3,145 @@
-      it "returns the constant" do
-        is_expected.to be MyAllocator
-      end
+      Object.send(:remove_const, :MyInventoryValidator)
```
```suggestion
  ensure
      Object.send(:remove_const, :MyInventoryValidator)
```

kennyadsl commented:
@waiting-for-dev can we merge this?

jarednorman commented:
Needs documentation update.

jarednorman commented:
I've opened up a change request in Git Book to explain the use of these endpoints.

jarednorman commented:
I've implemented @waiting-for-dev's remaining suggestion now.

waiting-for-dev reviewed (approved):
Thanks, @jarednorman! :heart:
